File: leetcode/082_removeduplicatesfromsortedlistII.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        
        if head is None or head.next is None:
            return head
        
        dummy = ListNode(0)
        dummy.next = head
        prev = dummy
        curr = head
        
        while curr is not None and curr.next is not None:
            if curr.val == curr.next.val:
                tmp = curr.next
                while tmp is not None and tmp.val == curr.val:
                    tmp = tmp.next
                prev.next = tmp
                curr = tmp
            else:
                curr = curr.next
                prev = prev.next
        return dummy.next

File: leetcode/082_removeduplicatesfromsortedlistII.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function deleteDuplicates(head: ListNode | null): ListNode | null {
    let left = new ListNode(0, head)
    let head2 = left
    let first = head
    let last = head

    while (last) {
        while (last && last.val === first.val) {
            last = last.next
        }
        if (first.next === last) { // no duplicates
            left = first
        } else {
            left.next = last
        }

        first = last
    }
    return head2.next
};

function printList(head: ListNode | null) {
    const result = []
    while (head) {
        result.push(head.val)
        head = head.next
    }
    console.log(result)
}
